### PR TITLE
Brand new reshape tool

### PR DESCRIPTION
**Pull request by nirvn**
This PR adds a new geometry editor for line and polygon layers, namely the reshape tool:
image
_The new tool's icon_

In action:
Peek 2020-12-25 16-17

It pretty much works like QGIS, no big behavior change. The tool will respect and handle topology (if the topology mode is toggled on) as well as intersection restraints set in the project.

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 0580d7ac93c62433d2ed48e2e94bf1bd22efc1ba:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-19b8983b23ddd72f67a36de35fdf5f62b9912be7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-19b8983b23ddd72f67a36de35fdf5f62b9912be7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-19b8983b23ddd72f67a36de35fdf5f62b9912be7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-19b8983b23ddd72f67a36de35fdf5f62b9912be7-x86.apk)

**Comment by suricactus** [comment]
I am looking at the gif and it seems the bottom right button is hidden, but after finishing it appears and fades away?

**Comment by nirvn** [comment]
@suricactus , that has to do with mouse emulating stylus & how we show/hide the (+) button when stylus hovers the screen.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 4703866a95c241b8b1d549db1bcc8b538cd95854:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-3bd1a967bb49353043ce93382eeaa8aa28963a66-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-3bd1a967bb49353043ce93382eeaa8aa28963a66-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-3bd1a967bb49353043ce93382eeaa8aa28963a66-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-3bd1a967bb49353043ce93382eeaa8aa28963a66-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 746da1c0a2e0ef027360314d2f370bdf89e97ef3:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-632442d1a95f84e19a6eeaf6d3f8ecf54018bee7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-632442d1a95f84e19a6eeaf6d3f8ecf54018bee7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-632442d1a95f84e19a6eeaf6d3f8ecf54018bee7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-632442d1a95f84e19a6eeaf6d3f8ecf54018bee7-x86.apk)

File: src/core/rubberbandmodel.cpp [hunk 1]
```diff
@@ -193,6 +196,9 @@ QgsPoint RubberbandModel::currentPoint( const QgsCoordinateReferenceSystem &crs,
 
 QgsPoint RubberbandModel::currentCoordinate() const
 {
+  if ( mPointList.isEmpty() )
+    return QgsPoint();
+
   return mPointList.at( mCurrentCoordinateIndex );
```
Comment (m-kuhn) [inline src/core/rubberbandmodel.cpp:202]:
```suggestion
  return mPointList.value( mCurrentCoordinateIndex );
```

that should protect for any index out of bounds.

File: src/core/utils/geometryutils.cpp [hunk 2]
```diff
@@ -38,6 +39,66 @@ QgsGeometry GeometryUtils::polygonFromRubberband( RubberbandModel *rubberBandMod
...
+        case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsCurrentLayer:
+          avoidIntersectionsLayers.append( layer );
+          break;
+        case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsLayers:
+          avoidIntersectionsLayers = QgsProject::instance()->avoidIntersectionsLayers();
+          break;
+        case QgsProject::AvoidIntersectionsMode::AllowIntersections:
+          break;
+      }
+      if ( avoidIntersectionsLayers.size() > 0 )
```
Comment (m-kuhn) [inline src/core/utils/geometryutils.cpp:70]:
```suggestion
      if ( !avoidIntersectionsLayers.empty() )
```

File: src/core/utils/geometryutils.cpp [hunk 3]
```diff
@@ -38,6 +39,66 @@ QgsGeometry GeometryUtils::polygonFromRubberband( RubberbandModel *rubberBandMod
   return g;
 }
 
+QgsGeometry::OperationResult GeometryUtils::reshapeFromRubberband( QgsVectorLayer *layer, QgsFeatureId fid, RubberbandModel *rubberBandModel )
+{
+  QgsFeature feature = layer->getFeature( fid );
+  if ( feature.geometry().isNull() ||
+       ( QgsWkbTypes::geometryType( layer->wkbType() ) != QgsWkbTypes::LineGeometry && QgsWkbTypes::geometryType( layer->wkbType() ) != QgsWkbTypes::PolygonGeometry ) )
```
Comment (m-kuhn) [inline src/core/utils/geometryutils.cpp:46]: Should we check layer of feature type here? I know this should currently match, but we never know for the future

File: src/core/utils/geometryutils.cpp [hunk 4]
```diff
@@ -38,6 +39,66 @@ QgsGeometry GeometryUtils::polygonFromRubberband( RubberbandModel *rubberBandMod
...
+    return QgsGeometry::InvalidBaseGeometry;
+
+  QgsPointSequence points = rubberBandModel->pointSequence( layer->crs(), QgsWkbTypes::Point, false );
+  QgsLineString reshapeLineString( points );
+
+  QgsGeometry geom = feature.geometry();
+  QgsGeometry::OperationResult reshapeReturn = geom.reshapeGeometry( reshapeLineString );
+  if ( reshapeReturn == QgsGeometry::Success ) {
```
Comment (m-kuhn) [inline src/core/utils/geometryutils.cpp:54]: Can you run astyle/prepare-commit.sh on this file?

File: src/core/utils/geometryutils.cpp [hunk 5]
```diff
@@ -38,6 +39,66 @@ QgsGeometry GeometryUtils::polygonFromRubberband( RubberbandModel *rubberBandMod
...
+          break;
+        case QgsProject::AvoidIntersectionsMode::AvoidIntersectionsLayers:
+          avoidIntersectionsLayers = QgsProject::instance()->avoidIntersectionsLayers();
+          break;
+        case QgsProject::AvoidIntersectionsMode::AllowIntersections:
+          break;
+      }
+      if ( avoidIntersectionsLayers.size() > 0 )
+      {
+        if ( geom.avoidIntersections( QgsProject::instance()->avoidIntersectionsLayers() ) != 0 )
```
Comment (m-kuhn) [inline src/core/utils/geometryutils.cpp:72]:
```suggestion
        if ( geom.avoidIntersections( avoidIntersectionsLayers ) != 0 )
```

no?

**Review by m-kuhn — APPROVED** [review]
Have a look at the comments and then it's good to go

**Comment by nirvn** [comment]
@m-kuhn , comments addressed.

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for d6fe8c1b2ed9e0e88c64f039427bc0f78285657a:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-c4e74315b536d1e7c057c000bfedb4c8a7d31dc3-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-c4e74315b536d1e7c057c000bfedb4c8a7d31dc3-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-c4e74315b536d1e7c057c000bfedb4c8a7d31dc3-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-c4e74315b536d1e7c057c000bfedb4c8a7d31dc3-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 737f1340645a31b30c9d876f7ce06f7675fc5f64:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-118653f409e3af6573c2c8ea348ee26cd20a87f7-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-118653f409e3af6573c2c8ea348ee26cd20a87f7-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-118653f409e3af6573c2c8ea348ee26cd20a87f7-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-118653f409e3af6573c2c8ea348ee26cd20a87f7-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 7c6f3eb6ccc0abbb6f8d8cb099d0cef18d65ea0d:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-b76bdf7b73d79518b3385a55575674cec25625e3-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-b76bdf7b73d79518b3385a55575674cec25625e3-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-b76bdf7b73d79518b3385a55575674cec25625e3-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-b76bdf7b73d79518b3385a55575674cec25625e3-x86.apk)

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for e2c3f0dab3981b53c3452e487071f74e87313722:
  - [**arm64_v8a**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-9b599d5c5c61ca8d310bbc19a67f53e83104653d-arm64_v8a.apk)

Other architectures: [armv7](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-9b599d5c5c61ca8d310bbc19a67f53e83104653d-armv7.apk), [x86_64](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-9b599d5c5c61ca8d310bbc19a67f53e83104653d-x86_64.apk), [x86](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-1527-9b599d5c5c61ca8d310bbc19a67f53e83104653d-x86.apk)

**Pull request merged**